### PR TITLE
Added caching: Twitterservice now only calls twitter for timeline/fil…

### PR DESCRIPTION
**Pull request by dhagandev**
…tered tweets if cache is expired or there are too many entries in cache

### CODE REVIEW

File: src/main/java/liuni/TimeCache.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+        public boolean outOfDate() {
+            Date now = new Date();
+            long different = now.getTime() - creationDate.getTime();
+            return different > timeToLive;
+        }
+    }
+
+    public TimeCache(long maxTimeSec, int maxItems) {
+        maxEntries = maxItems;
+        timeToLive = maxTimeSec * 1000;
```
Comment (kev-chan) [inline src/main/java/liuni/TimeCache.java]: use constant vs hardcoded number

File: src/main/java/liuni/services/TwitterService.java [hunk 2]
```diff
@@ -39,26 +43,30 @@ public Twitter getTwitter() {
...
         }
         return Optional.empty();
     }
 
     public Optional<List<TwitterTweetModel>> getTimeline() throws TwitterException {
-        return Optional.of(twitter.getHomeTimeline().stream()
-                                  .map(status -> getTweet(status))
-                                  .collect(Collectors.toList()));
+        String cacheKey = "timeline_" + new Date();
+        return (cache.getEntry(cacheKey) != null) ? Optional.of(cache.getEntry(cacheKey)) : Optional.of(cache.putEntry(cacheKey, twitter.getHomeTimeline().stream()
```
Comment (kev-chan) [inline src/main/java/liuni/services/TwitterService.java]: why the addition of the date?

**Review by kev-chan — CHANGES_REQUESTED** [review]

File: src/main/java/liuni/services/TwitterService.java [hunk 3]
```diff
@@ -39,26 +43,30 @@ public Twitter getTwitter() {
...
+                         .findFirst();
         }
         return Optional.empty();
     }
 
     public Optional<List<TwitterTweetModel>> getTimeline() throws TwitterException {
-        return Optional.of(twitter.getHomeTimeline().stream()
-                                  .map(status -> getTweet(status))
-                                  .collect(Collectors.toList()));
+        String cacheKey = "timeline";
```
Comment (kev-chan) [inline src/main/java/liuni/services/TwitterService.java]: make a constant

File: src/main/java/liuni/services/TwitterService.java [hunk 4]
```diff
@@ -39,26 +43,30 @@ public Twitter getTwitter() {
...
+                                                                                                                                        .map(status -> getTweet(status))
+                                                                                                                                        .collect(Collectors.toList())));
     }
 
     public Optional<List<TwitterTweetModel>> getFiltered(String filterKey) throws TwitterException {
-        return Optional.of(twitter.getHomeTimeline().stream()
-                                  .filter(status -> status.getText().toLowerCase().contains(filterKey.toLowerCase()))
-                                  .map(status -> getTweet(status))
-                                  .collect(Collectors.toList()));
+        String cacheKey = "filter_" + filterKey;
```
Comment (kev-chan) [inline src/main/java/liuni/services/TwitterService.java]: make prefix a constant

File: src/main/java/liuni/TimeCache.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    public List<TwitterTweetModel> getEntry(String key) {
+        removeOldEntries();
+        if (isEntry(key)) {
+            return cache.get(key).data;
+        }
+        return null;
+    }
+
+    public List<TwitterTweetModel> putEntry(String key, List<TwitterTweetModel> value) {
+        if (key.equals("timeline")) {
```
Comment (kev-chan) [inline src/main/java/liuni/TimeCache.java]: why the need to distinguish between timeline?

**Review by kev-chan — CHANGES_REQUESTED** [review]

File: src/main/java/liuni/TimeCache.java (hunk 5 above)
Comment (dhagandev) [reply src/main/java/liuni/TimeCache.java]: To list the timeline entry as important so that it is not removed if you try to add more entries to the cache

Comment (kev-chan) [reply src/main/java/liuni/TimeCache.java]: is it not the case where the oldest value gets removed anyway?

**Pull request merged**